service/raft/raft_group_registry.cc

```diff
@@ -123,13 +125,30 @@ class gossiper_state_change_subscriber_proxy: public gms::i_endpoint_state_chang
 
     virtual future<>
     on_alive(gms::inet_address endpoint, gms::endpoint_state_ptr ep_state, gms::permit_id) override {
-        co_await utils::get_local_injector().inject_with_handler("raft_group_registry::on_alive", [endpoint, ep_state] (auto& handler) -> future<> {
-            auto app_state_ptr = ep_state->get_application_state_ptr(gms::application_state::HOST_ID);
-            if (!app_state_ptr) {
-                co_return;
-            }
-            
-            raft::server_id id(utils::UUID(app_state_ptr->value()));
+        auto app_state_ptr = ep_state->get_application_state_ptr(gms::application_state::HOST_ID);
+        if (!app_state_ptr) {
+            return make_ready_future();
+        }
+        raft::server_id id(utils::UUID(app_state_ptr->value()));
+        auto generation = ep_state->get_heart_beat_state().get_generation();
+
+#ifndef SCYLLA_ENABLE_ERROR_INJECTION
+        on_endpoint_change(endpoint, id, generation);
+        return make_ready_future();
+#else
+        // wait in the background since this notification callabck is called
+        // with the endpoint locked
+        // It is safe to discard this future as the gate is closed in drain_on_shutdown()
+        (void)with_gate(_registry.async_gate(), [this, endpoint, id, generation] {
```

Comment:
I think it might be better to move the `inject_with_handler` into `real_mark_alive` call site:
  * no need to complicate the matters here (`async_gate`, separate code paths for injection and non injection), simple `inject_with_handler` should continue to work in `gossiper` just like it used to work here;
  * the injection would be 'closer to the real world' there - it would imitate the network delay in getting the echo message reply from an endpoint;
  * as it would become more general, it can be reused in other tests

Reply:
makes sense